{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "100b5866-d605-44a2-b9fc-b6c9b62492e5",
   "metadata": {},
   "source": [
    "This jupyter notebook contains the supplementary Python Code used in the manuscript **G4-quadruplex-mediated genomic instability drives SNVs in cancer** by Richl et al.\n",
    "\n",
    "# Contents\n",
    "\n",
    "- 1. Libraries\n",
    "- 2. Gathering datasets\n",
    "- 3. Analyzing and plotting the spatial distribution of cSNVs or SNPs at G4s.\n",
    "- 4. Analyzing the density distribution of G4s and cSNVs in the CDK4 gene.\n",
    "- 5. Analyzing and plotting the average number of cSNVs or random SNPs in genomic 5 kbps windows containing either 0/!0/>=10 G4s.\n",
    "- 6. Stratifying genomic 5 kbps windows based on the number of G4s contained and plotting the distribution of cSNVs or SNPs.\n",
    "- 7. Analyzing and plotting the average number of G4s in non-cancer or cancer genes.\n",
    "- 8. Analyzing the density distribution of cSNVs or SNPs in either cancer- or non-cancer genes for genetic windows with 0 / >=10 G4s.\n",
    "- 9. Perform incomplete categorical bootstrapping and approximate the limit R2 for increasing random sampling sizes.\n",
    "- 10. Plot incomplete categorical bootstrapping for the full dataset.\n",
    "- 11. Comparing the average number of normalized cSNVs across different cancer types in genomic windows containing 0/>=10 G4s.\n",
    "- 12. Group patients by cancer type and tissue and calculate median G4-averages across patients with similiar diagnoses. Plot as sunburst plot.\n",
    "- 13. Group patients by cancer type and calculate intra-group mean-normalized standard deviation.\n",
    "- 14. Fit Cox's proportional hazards model to patients that are eeither alive or died due to cancer-related circumstances. \n",
    "- 15. Plot results of Cox's proportional hazards model.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "759036ba-78b2-4dd5-9c85-0a4548f04357",
   "metadata": {},
   "source": [
    "# 1. Libraries\n",
    "The following libraries are used throughout the notebook and need to be installed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "535014f9-44ac-4b22-842b-ceb2dd6fe904",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.linear_model import LinearRegression\n",
    "import seaborn as sns\n",
    "from sklearn import linear_model\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import plotly.io as pio\n",
    "import plotly.express as px\n",
    "import matplotlib.pyplot as plt\n",
    "import plotly.figure_factory as ff\n",
    "import plotly.graph_objects as go\n",
    "from scipy.stats import gaussian_kde\n",
    "import matplotlib.ticker as ticker\n",
    "from lifelines import CoxPHFitter\n",
    "from matplotlib.pyplot import figure\n",
    "\n",
    "\n",
    "# Setting global options \n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.max_rows', 500)\n",
    "sns.set_theme()\n",
    "\n",
    "# Set colors\n",
    "cSNPs = \"#e69422\"\n",
    "SNPs = \"#1e60c9\"\n",
    "Random = \"#000000\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "822a934a-f4a7-4db4-acd7-46050f69d1d6",
   "metadata": {},
   "source": [
    "# 2. Gathering Datasets\n",
    "Loading the data and building one final dataframe holding all cSNVs with associated meta- and window-data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77b42c74-1164-464e-97b4-b1e73fc63176",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "### Loading the Dataframe with all Mutations\n",
    "\n",
    "df = pd.read_csv(\"~/SNP_exons\", sep=\"\\t\", header=None).drop(columns={5,6,7,8,9,10}).rename(columns={0: \"SNP_ID\", \n",
    "                                                                                              1: \"Strand\", \n",
    "                                                                                              2: \"Type\", \n",
    "                                                                                              3: \"Reference_Allele\", \n",
    "                                                                                              4: \"Allele\", \n",
    "                                                                                              11: \"case_id\",\n",
    "                                                                                              12: \"window_id\"}).drop_duplicates(keep=\"first\")\n",
    "df2 = pd.read_csv(\"~/exons_G4.count\", sep=\"\\t\", header=None).drop(columns={0,1,2}).rename(columns={3: \"window_id\", 4: \"G4_Count\"})\n",
    "df3 = pd.read_csv(\"~/exons_SNP.count\", sep=\"\\t\", header=None).drop(columns={0,1,2}).rename(columns={3: \"window_id\", 4: \"SNP_Count\"})\n",
    "\n",
    "df_m1 = pd.merge(df2,df, on=\"window_id\", how=\"left\")#.fillna(0)\n",
    "df_m1[\"G4_Count\"] = df_m1[\"G4_Count\"].fillna(0)\n",
    "df_m1[\"SNP_ID\"] = df_m1[\"SNP_ID\"].fillna(0)\n",
    "df_m2 = pd.merge(df_m1,df3, on=\"window_id\", how=\"left\")#.fillna(0)\n",
    "df_c = pd.read_csv(\"~/clinical.tsv\", sep=\"\\t\")\n",
    "df_c2 = df_c[[\"case_id\", \"cause_of_death\", \"days_to_birth\", \"primary_diagnosis\", \"therapeutic_agents\", \"vital_status\", \"tissue_or_organ_of_origin\"]].drop_duplicates(subset=\"case_id\", keep=\"first\")\n",
    "df_m3 = pd.merge(df_m2, df_c2, on=\"case_id\", how=\"left\")\n",
    "\n",
    "drop_list = [\"case_id\", \"ajcc_clinical_t\", \"ajcc_clinical_n\", \"ajcc_clinical_m\", \"ajcc_pathologic_stage\", \"ajcc_pathologic_t\", \"ajcc_pathologic_n\", \"ajcc_pathologic_m\", \"prior_malignancy\"]\n",
    "df_test = pd.merge(df_m3, df_c[drop_list], on = \"case_id\", how=\"left\")\n",
    "df_m3 = df_test\n",
    "\n",
    "### Generating the Dataframe with cSNPs grouped by windows\n",
    "\n",
    "final = df_m3.groupby([\"window_id\"]).mean()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c1b6295-00c2-4c00-9aef-c5e5c43ea1cb",
   "metadata": {},
   "source": [
    "## Analyzing and plotting the spatial distribution of cSNVs or SNPs at G4s."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5cde03d-68f8-4ca1-8a02-c189ec91a69d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data Loading\n",
    "df1 = pd.read_csv(\"~/Common_SNPs_exons_5kb_windows_G4s.Distance\", delim_whitespace=True, header=None)[0].value_counts().reset_index().rename(columns={0: \"Count\", \"index\": \"Distance\"})\n",
    "df2 = pd.read_csv(\"~/random_exons_5kb_windows_G4s.Distance\", delim_whitespace=True, header=None)[0].value_counts().reset_index().rename(columns={0: \"Count\", \"index\": \"Distance\"})\n",
    "df3 = pd.read_csv(\"~/SNPs_exons_5kb_windows_G4s.Distance\", delim_whitespace=True, header=None)[0].value_counts().reset_index().rename(columns={0: \"Count\", \"index\": \"Distance\"})\n",
    "df1[\"Origin\"] = \"SNP\"\n",
    "df2[\"Origin\"] = \"Random\"\n",
    "df3[\"Origin\"] = \"cSNV\"\n",
    "df4 = pd.concat([df2,df1,df3])\n",
    "\n",
    "# Colors and group labels\n",
    "colors = [cSNPs, Random, SNPs]\n",
    "group_labels = [\"cSNV\", \"Random\", \"SNP\"]\n",
    "\n",
    "# Plotting with Matplotlib\n",
    "fig, ax = plt.subplots(figsize=(8, 8))\n",
    "for label, color in zip(group_labels, colors):\n",
    "    data = df4[df4[\"Origin\"] == label][\"Distance\"]\n",
    "    density = gaussian_kde(data)\n",
    "    xs = np.linspace(-200000, 200000, 1000)\n",
    "    density.covariance_factor = lambda: 0.25\n",
    "    density._compute_covariance()\n",
    "    ax.plot(xs, density(xs), label=label, color=color)\n",
    "\n",
    "ax.set_xlim(-200000, 200000)\n",
    "ax.set_xlabel(\"Distance of SNV to closest G4 in bp\")\n",
    "ax.set_ylabel(\"Normalized Density\")\n",
    "ax.legend(fontsize=14)\n",
    "\n",
    "# Format x-axis labels and add minor ticks\n",
    "ax.xaxis.set_major_formatter(ticker.FuncFormatter(lambda x, p: format(int(x/1000), ',')))\n",
    "ax.xaxis.set_minor_locator(ticker.AutoMinorLocator())\n",
    "ax.tick_params(which='both', width=1)\n",
    "ax.tick_params(which='major', length=6)\n",
    "ax.tick_params(which='minor', length=3)\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b30fb2a4-7efb-4d0c-996f-b6f2a6f6a42b",
   "metadata": {},
   "source": [
    "## Analyzing the density distribution of G4s and cSNVs in the *CDK4* gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5f90947-120a-4733-9960-08d411058dcf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Data Loading\n",
    "\n",
    "df_G4 = pd.read_csv(\"CDK4.G4s\", sep=\"\\t\", header=None).rename(columns={0: \"Chr\", 1: \"Start\", 2: \"Stop\", 3: \"ID\"})\n",
    "df_GDC = pd.read_csv(\"CDK4.ALL_GDC\", sep=\"\\t\", header=None)[[0,1,2,3]].rename(columns={0: \"Chr\", 1: \"Start\", 2: \"Stop\", 3: \"ID\"})\n",
    "df_G4[\"Class\"] = \"G4\"\n",
    "df_GDC[\"Class\"] = \"GDC\"\n",
    "df_all = pd.concat([df_G4, df_GDC])\n",
    "df_all[\"Summit\"] = (df_all[\"Stop\"] - df_all[\"Start\"]) / 2 + df_all[\"Start\"]\n",
    "\n",
    "# Plotting\n",
    "\n",
    "sns.set(rc={'figure.figsize':(10, 4)})\n",
    "\n",
    "sns.set(font_scale=2)\n",
    "sns.set_style(\"white\")\n",
    "ax = sns.kdeplot(data=df_all[df_all[\"Class\"] == \"G4\"][\"Summit\"], color = \"#000008\", bw_adjust=0.15)\n",
    "sns.kdeplot(data=df_all[df_all[\"Class\"] == \"GDC\"][\"Summit\"], bw_adjust=0.4, color = cSNPs)\n",
    "sns.despine()\n",
    "plt.savefig(\"output_seaborn_plot.png\", dpi=300)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c40e7e4e-624c-45d1-9016-df9c9cce1021",
   "metadata": {},
   "source": [
    "## Analyzing and plotting the average number of cSNVs or random SNPs in genomic 5 kbps windows containing either 0/!0/>=10 G4s."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45335419-4ab6-45f7-aa20-06483b8b38af",
   "metadata": {},
   "outputs": [],
   "source": [
    "lower_limit = 1\n",
    "test1 = final[final[\"G4_Count\"] == 0][\"SNP_Count\"].sample(5000).clip(lower=lower_limit)\n",
    "test2 = final[final[\"G4_Count\"] != 0][\"SNP_Count\"].sample(5000).clip(lower=lower_limit)\n",
    "test3 = final[final[\"G4_Count\"] > 9][\"SNP_Count\"].clip(lower=lower_limit)\n",
    "test4 = final_random[final_random[\"G4_Count\"] == 0][\"SNP_Count\"].sample(5000).clip(lower=lower_limit)\n",
    "test5 = final_random[final_random[\"G4_Count\"] != 0][\"SNP_Count\"].sample(5000).clip(lower=lower_limit)\n",
    "test6 = final_random[final_random[\"G4_Count\"] > 9][\"SNP_Count\"].clip(lower=lower_limit)\n",
    "fig = go.Figure()\n",
    "\n",
    "# Define a helper function to add box plots with adjusted x values\n",
    "def add_box_trace(fig, y, name, fillcolor, x_offset):\n",
    "    fig.add_trace(go.Box(\n",
    "        y=y,\n",
    "        name=name,\n",
    "        boxmean=True,\n",
    "        boxpoints=False,\n",
    "        line_color='black',\n",
    "        fillcolor=fillcolor,\n",
    "        opacity=0.6,\n",
    "        x=[x_offset] * len(y)\n",
    "    ))\n",
    "\n",
    "\n",
    "add_box_trace(fig, test1, 'No G4-Hit', 'lightsteelblue', 1)\n",
    "add_box_trace(fig, test2, 'G4-Hit', 'lightcoral', 2)\n",
    "add_box_trace(fig, test3, '>= 10 G4s', 'lightseagreen', 3)\n",
    "\n",
    "\n",
    "add_box_trace(fig, test4, 'No G4-Hit', 'lightsteelblue', 4)\n",
    "add_box_trace(fig, test5, 'G4-Hit', 'lightcoral', 5)\n",
    "add_box_trace(fig, test6, '>= 10 G4s', 'lightseagreen', 6)\n",
    "\n",
    "# Update the layout\n",
    "fig.update_layout(\n",
    "    autosize=False,\n",
    "    width=600,\n",
    "    height=600,\n",
    "    margin=dict(\n",
    "        l=50,\n",
    "        r=50,\n",
    "        b=100,\n",
    "        t=100,\n",
    "        pad=4\n",
    "    ),\n",
    "    yaxis_title=\"SNPs / 5 kpbs\",\n",
    "    xaxis_title=\"Groups\",\n",
    "    xaxis=dict(\n",
    "        tickvals=[1.5, 4.5],\n",
    "        ticktext=[\"Samples\", \"Control\"],\n",
    "        tickfont=dict(size=25)\n",
    "    ),\n",
    "    yaxis=dict(tickfont=dict(size=25)),\n",
    "    template=\"simple_white\",\n",
    "    font=dict(size=25),\n",
    "    showlegend=False\n",
    ")\n",
    "\n",
    "# Set y-axis to logarithmic scale\n",
    "fig.update_yaxes(type=\"log\")\n",
    "\n",
    "# Show the plot\n",
    "fig.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9558213b-ac31-4aaf-8ad9-a8893b62c68c",
   "metadata": {},
   "source": [
    "## Stratifying genomic 5 kbps windows based on the number of G4s contained and plotting the distribution of cSNVs or SNPs. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bd88531-8fc5-43a7-bf2c-2e13cca5c779",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = []\n",
    "fontsize=15\n",
    "for G4 in range(0,16,1):\n",
    "    data.append(list(df[df[\"G4_Count\"] == G4][\"cSNP_Count_normalized\"]))\n",
    "fig, ax = plt.subplots(figsize=(10,10))\n",
    "parts = plt.violinplot(data,quantiles = None,\n",
    "                        showmeans=False, \n",
    "                        showmedians=False,\n",
    "                        showextrema=False,\n",
    "                        widths = 0.8)\n",
    "for pc in parts['bodies']:\n",
    "    pc.set_facecolor(cSNPs)\n",
    "    pc.set_edgecolor('black')\n",
    "    pc.set_alpha(0.5)\n",
    "\n",
    "x1 = df[df[\"G4_Count\"] <= 18].groupby([\"G4_Count\"]).mean().reset_index()[\"G4_Count\"].values\n",
    "y1 = df[df[\"G4_Count\"] <= 18].groupby([\"G4_Count\"]).mean().reset_index()[\"cSNP_Count_normalized\"].values\n",
    "\n",
    "model1 = np.poly1d(np.polyfit(x1, y1, 1))\n",
    "\n",
    "line1 = np.linspace(min(x1), max(x1), 1000)\n",
    "ax.plot(line1, model1(line1), color=\"black\", lw = 5)\n",
    "\n",
    "ax.scatter(df[(df[\"G4_Count\"] < 18) & (df[\"G4_Count\"] > 0)].groupby([\"G4_Count\"]).mean().reset_index()[\"G4_Count\"], df[(df[\"G4_Count\"] < 18) & (df[\"G4_Count\"] > 0)].groupby([\"G4_Count\"]).mean().reset_index()[\"cSNP_Count_normalized\"], color=\"black\", marker=\"+\", s=100,zorder=2)\n",
    "ax.set_title('cSNPs')\n",
    "for pos in ['right', 'top']:\n",
    "    plt.gca().spines[pos].set_visible(False)\n",
    "plt.rcParams.update({\n",
    "    'font.size': fontsize,         # default font size\n",
    "    'axes.labelsize': fontsize,    # x and y labels\n",
    "    'xtick.labelsize': fontsize,   # x-axis tick labels\n",
    "    'ytick.labelsize': fontsize,   # y-axis tick labels\n",
    "    'legend.fontsize': fontsize,   # legend font size\n",
    "    'figure.titlesize': fontsize,  # figure title font size\n",
    "})\n",
    "plt.gca().set_yscale('log')\n",
    "ax.set_ylabel('Normalized SNPs / 5kb')\n",
    "ax.set_xlabel('G4s / 5kb')\n",
    "plt.ylim([0.1, 100])\n",
    "plt.savefig(\"output_stratify_1_plot.png\", dpi=500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dae655a9-b7e6-4f97-9d21-e40437aa5e01",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = []\n",
    "fontsize=5\n",
    "for G4 in range(0,16,1):\n",
    "    data.append(list(df[df[\"G4_Count\"] == G4][\"SNP_Count_normalized\"]))\n",
    "fig, ax = plt.subplots(figsize=(10,10))\n",
    "parts = plt.violinplot(data,quantiles = None,\n",
    "                        showmeans=False, \n",
    "                        showmedians=False,\n",
    "                        showextrema=False,\n",
    "                        widths = 0.8)\n",
    "for pc in parts['bodies']:\n",
    "    pc.set_facecolor(SNPs)\n",
    "    pc.set_edgecolor('black')\n",
    "    pc.set_alpha(0.5)\n",
    "\n",
    "x1 = df[df[\"G4_Count\"] <= 18].groupby([\"G4_Count\"]).mean().reset_index()[\"G4_Count\"].values\n",
    "y1 = df[df[\"G4_Count\"] <= 18].groupby([\"G4_Count\"]).mean().reset_index()[\"SNP_Count_normalized\"].values\n",
    "\n",
    "model1 = np.poly1d(np.polyfit(x1, y1, 1))\n",
    "\n",
    "line1 = np.linspace(min(x1), max(x1), 1000)\n",
    "ax.plot(line1, model1(line1), color=\"black\", lw = 5)\n",
    "\n",
    "ax.scatter(df[(df[\"G4_Count\"] < 18) & (df[\"G4_Count\"] > 0)].groupby([\"G4_Count\"]).mean().reset_index()[\"G4_Count\"], df[(df[\"G4_Count\"] < 18) & (df[\"G4_Count\"] > 0)].groupby([\"G4_Count\"]).mean().reset_index()[\"SNP_Count_normalized\"], color=\"black\", marker=\"+\", s=100,zorder=2)\n",
    "ax.set_title('SNPs')\n",
    "for pos in ['right', 'top']:\n",
    "    plt.gca().spines[pos].set_visible(False)\n",
    "plt.rcParams.update({\n",
    "    'font.size': fontsize,         # default font size\n",
    "    'axes.labelsize': fontsize,    # x and y labels\n",
    "    'xtick.labelsize': fontsize,   # x-axis tick labels\n",
    "    'ytick.labelsize': fontsize,   # y-axis tick labels\n",
    "    'legend.fontsize': fontsize,   # legend font size\n",
    "    'figure.titlesize': fontsize,  # figure title font size\n",
    "})\n",
    "plt.gca().set_yscale('log')\n",
    "ax.set_xlabel('G4s / 5kb')\n",
    "plt.ylim([0.1, 100])\n",
    "plt.savefig(\"output_stratify_2_plot.png\", dpi=500)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3e46e44-299c-4484-b284-3d5df483665a",
   "metadata": {},
   "source": [
    "## Analyzing and plotting the average number of G4s in non-cancer or cancer genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3973ecc-fd3a-4f62-845b-c4ba7dbc945e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Data\n",
    "\n",
    "df = pd.read_csv(\"genes_G4.count\", delim_whitespace=True, header=None).rename(columns={1: \"Gene\", 0: \"G4_Count\"})\n",
    "df_cg = pd.read_csv(\"Cancer_Genes\", sep=\"\\t\").rename(columns={\"Gene Symbol\": \"Gene\"})\n",
    "df_cg[\"Cancer_Gene\"] = 1\n",
    "df_m = pd.merge(df,df_cg, on=\"Gene\", how=\"left\").fillna(0).drop_duplicates(subset=\"Gene\", keep=\"first\")\n",
    "\n",
    "# Plotting\n",
    "\n",
    "test1 = df_m[df_m[\"Cancer_Gene\"] == 0][\"G4_Count\"]\n",
    "test2 = df_m[df_m[\"Cancer_Gene\"] != 0][\"G4_Count\"]\n",
    "fig = go.Figure()\n",
    "fig.add_trace(go.Box(\n",
    "    y=test1,\n",
    "    name='Non-cancer gene',\n",
    "    #marker_color='#FF4136',\n",
    "    boxmean=True,\n",
    "    boxpoints=False,\n",
    "    line_color='black',\n",
    "    fillcolor=\"lightsteelblue\",\n",
    "    opacity=0.6,\n",
    "))\n",
    "fig.add_trace(go.Box(\n",
    "    y=test2,\n",
    "    name='Cancer gene',\n",
    "    #marker_color='#FF851B',\n",
    "    boxmean=True,\n",
    "    boxpoints=False,\n",
    "    line_color='black',\n",
    "    fillcolor=\"lightcoral\",\n",
    "    opacity=0.6,\n",
    "))\n",
    "\n",
    "fig.update_layout(\n",
    "    autosize=False,\n",
    "    width=600,\n",
    "    height=600,\n",
    "    margin=dict(\n",
    "        l=50,\n",
    "        r=50,\n",
    "        b=100,\n",
    "        t=100,\n",
    "        pad=4\n",
    "    ),\n",
    ")\n",
    "\n",
    "#fig.update_yaxes(title_text=\"y-axis in logarithmic scale\", type=\"log\")\n",
    "\n",
    "fig.for_each_trace(lambda t: t.update({\"marker\":{\"size\":3}}))\n",
    "fig.update_yaxes(type=\"log\")\n",
    "fig.update_yaxes(title=\"log G4s / gene\")\n",
    "fig.update_layout(\n",
    "    template=\"simple_white\",\n",
    "    #legend_title=\"Legend Title\",\n",
    "    yaxis = dict(tickfont = dict(size=25)),\n",
    "    xaxis = dict(tickfont = dict(size=25)),\n",
    "    #legend = dict(font = dict(size = 20, color = \"black\")),\n",
    "    font=dict(size=30))\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "335beca5-c9d3-48e9-ad83-a8acf287011d",
   "metadata": {},
   "source": [
    "## Analyzing the density distribution of cSNVs or SNPs in either cancer- or non-cancer genes for genetic windows with 0 / >=10 G4s."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ffaafad-e672-48cf-a68e-ac8c37c7637e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Data Loading\n",
    "\n",
    "df_GDC_cancer = pd.read_csv(\"ALL_GDC.sorted.bed.cancer.count\", header=None, delim_whitespace=True).rename(columns={1:\"window_id\", 0: \"GDC_Count\"})\n",
    "df_GDC_non_cancer = pd.read_csv(\"ALL_GDC.sorted.bed.non_cancer.count\", header=None, delim_whitespace=True).rename(columns={1:\"window_id\", 0: \"GDC_Count\"})\n",
    "df_G4 = pd.read_csv(\"sliding_5kb_windows_hg38_exons.G4_count\", header=None, sep=\"\\t\").drop(columns={0,1,2}).rename(columns={3: \"window_id\", 4: \"G4_Count\"})\n",
    "df_m_GDC_cancer = pd.merge(df_GDC_cancer, df_G4, on=\"window_id\", how=\"left\")\n",
    "df_m_GDC_non_cancer = pd.merge(df_GDC_non_cancer, df_G4, on=\"window_id\", how=\"left\")\n",
    "\n",
    "\n",
    "df_m_GDC_cancer[\"log_GDC_Count\"] = np.log(df_m_GDC_cancer[\"GDC_Count\"])\n",
    "df_m_GDC_non_cancer[\"log_GDC_Count\"] = np.log(df_m_GDC_non_cancer[\"GDC_Count\"])\n",
    "test1 = df_m_GDC_cancer[df_m_GDC_cancer[\"G4_Count\"] == 0][\"log_GDC_Count\"]#.sample(10000)\n",
    "test2 = df_m_GDC_cancer[df_m_GDC_cancer[\"G4_Count\"] > 10][\"log_GDC_Count\"]\n",
    "test3 = df_m_GDC_non_cancer[df_m_GDC_non_cancer[\"G4_Count\"] == 0][\"log_GDC_Count\"]\n",
    "test4 = df_m_GDC_non_cancer[df_m_GDC_non_cancer[\"G4_Count\"] > 10][\"log_GDC_Count\"]\n",
    "\n",
    "fig = go.Figure()\n",
    "fig.add_trace(go.Violin(\n",
    "    x=test1,\n",
    "    name='',\n",
    "    marker_color=cSNPs,\n",
    "    #fillcolor=\"white\",\n",
    "    spanmode=\"hard\",\n",
    "    side='negative',\n",
    "    #opacity=1,\n",
    "    points=False,\n",
    "    \n",
    "))\n",
    "fig.add_trace(go.Violin(\n",
    "    x=test2,\n",
    "    name='',\n",
    "    marker_color='#9e6516',\n",
    "    #fillcolor=\"white\",\n",
    "    spanmode=\"hard\",\n",
    "    side='negative',\n",
    "    #opacity=1,\n",
    "    points=False\n",
    "))\n",
    "\n",
    "fig.add_trace(go.Violin(\n",
    "    x=test3,\n",
    "    name='',\n",
    "    marker_color=cSNPs,\n",
    "    #fillcolor=\"white\",\n",
    "    spanmode=\"hard\",\n",
    "    side='positive',\n",
    "    #opacity=0.1,\n",
    "    points=False\n",
    "))\n",
    "fig.add_trace(go.Violin(\n",
    "    x=test4,\n",
    "    name='',\n",
    "    marker_color='#9e6516',\n",
    "    #fillcolor=\"white\",\n",
    "    spanmode=\"hard\",\n",
    "    side='positive',\n",
    "    #opacity=1,\n",
    "    points=False\n",
    "))\n",
    "\n",
    "fig.update_layout(\n",
    "    #title=\"Plot Title\",\n",
    "    yaxis_title=\"Cancer Genes          Normal Genes\",\n",
    "    #xaxis_title=\"Log SNPs / 5kb\",\n",
    ")\n",
    "\n",
    "fig.update_traces(box_visible=False, meanline_visible=True)\n",
    "#fig.update_layout(violinmode='group')\n",
    "fig.update_layout(\n",
    "    autosize=False,\n",
    "    width=600,\n",
    "    height=600,\n",
    ")\n",
    "fig.update_layout(\n",
    "    template=\"simple_white\",\n",
    "    #legend_title=\"Legend Title\",\n",
    "    yaxis = dict(tickfont = dict(size=15)),\n",
    "    xaxis = dict(tickfont = dict(size=20)),\n",
    "    #legend = dict(font = dict(size = 20, color = \"black\")),\n",
    "    font=dict(size=25))\n",
    "fig.update_layout(showlegend=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8edc57dd-b6b0-49b4-8191-8e47f3af3e5a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_SNP_cancer = pd.read_csv(\"Common_SNPs.bed.cancer.count\", header=None, delim_whitespace=True).rename(columns={1:\"window_id\", 0: \"SNP_Count\"})\n",
    "df_SNP_non_cancer = pd.read_csv(\"Common_SNPs.bed.non_cancer.count\", header=None, delim_whitespace=True).rename(columns={1:\"window_id\", 0: \"SNP_Count\"})\n",
    "df_G4 = pd.read_csv(\"sliding_5kb_windows_hg38_exons.G4_count\", header=None, sep=\"\\t\").drop(columns={0,1,2}).rename(columns={3: \"window_id\", 4: \"G4_Count\"})\n",
    "df_m_SNP_cancer = pd.merge(df_SNP_cancer, df_G4, on=\"window_id\", how=\"left\")\n",
    "df_m_SNP_non_cancer = pd.merge(df_SNP_non_cancer, df_G4, on=\"window_id\", how=\"left\")\n",
    "\n",
    "\n",
    "df_m_SNP_cancer[\"log_SNP_Count\"] = np.log(df_m_SNP_cancer[\"SNP_Count\"])\n",
    "df_m_SNP_non_cancer[\"log_SNP_Count\"] = np.log(df_m_SNP_non_cancer[\"SNP_Count\"])\n",
    "test1 = df_m_SNP_cancer[df_m_SNP_cancer[\"G4_Count\"] == 0][\"log_SNP_Count\"]#.sample(10000)\n",
    "test2 = df_m_SNP_cancer[df_m_SNP_cancer[\"G4_Count\"] > 10][\"log_SNP_Count\"]\n",
    "test3 = df_m_SNP_non_cancer[df_m_SNP_non_cancer[\"G4_Count\"] == 0][\"log_SNP_Count\"]\n",
    "test4 = df_m_SNP_non_cancer[df_m_SNP_non_cancer[\"G4_Count\"] > 10][\"log_SNP_Count\"]\n",
    "\n",
    "fig = go.Figure()\n",
    "fig.add_trace(go.Violin(\n",
    "    x=test1,\n",
    "    name='',\n",
    "    marker_color=SNPs,\n",
    "    spanmode=\"hard\",\n",
    "    side='negative',\n",
    "    points=False\n",
    "))\n",
    "fig.add_trace(go.Violin(\n",
    "    x=test2,\n",
    "    name='',\n",
    "    marker_color='#123c80',\n",
    "    spanmode=\"hard\",\n",
    "    side='negative',\n",
    "    points=False\n",
    "))\n",
    "fig.add_trace(go.Violin(\n",
    "    x=test3,\n",
    "    name='',\n",
    "    marker_color=SNPs,\n",
    "    spanmode=\"hard\",\n",
    "    side='positive',\n",
    "    points=False\n",
    "))\n",
    "fig.add_trace(go.Violin(\n",
    "    x=test4,\n",
    "    name='',\n",
    "    marker_color='#123c80',\n",
    "    spanmode=\"hard\",\n",
    "    side='positive',\n",
    "    points=False\n",
    "))\n",
    "\n",
    "fig.update_layout(\n",
    "    #title=\"SNPs\",\n",
    "    yaxis_title=\"Cancer Genes          Normal Genes\",\n",
    "    #xaxis_title=\"Log SNPs / 5kb\",\n",
    ")\n",
    "\n",
    "fig.update_traces(box_visible=False, meanline_visible=True)\n",
    "#fig.update_layout(violinmode='group')\n",
    "fig.update_layout(\n",
    "    autosize=False,\n",
    "    width=600,\n",
    "    height=600,\n",
    ")\n",
    "fig.update_layout(\n",
    "    template=\"simple_white\",\n",
    "    #legend_title=\"Legend Title\",\n",
    "    yaxis = dict(tickfont = dict(size=25)),\n",
    "    xaxis = dict(tickfont = dict(size=20)),\n",
    "    #legend = dict(font = dict(size = 20, color = \"black\")),\n",
    "    font=dict(size=30))\n",
    "fig.update_layout(showlegend=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed21c563-7eca-4599-9e57-e75a62223044",
   "metadata": {},
   "source": [
    "## Perform incomplete categorical bootstrapping and approximate the limit R2 for increasing random sampling sizes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "361ade70-0924-4797-b513-07f42ed0cbc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "key1 = \"SNP_Count\"\n",
    "key2 = \"cSNP_Count\"\n",
    "df_results = pd.DataFrame({\"Sample_Size\": [],\n",
    "                           \"R2_1\": [],\n",
    "                           \"R2_2\": [],\n",
    "                           \"Coef_1\": [],\n",
    "                           \"Coef_2\": []})\n",
    "\n",
    "# Incomplete Categorical Bootsrapping\n",
    "start = 0\n",
    "stop = 16\n",
    "inc = 1\n",
    "tolerance = 0.3\n",
    "permutation = 200\n",
    "for sample in range(1,101,5):\n",
    "    df_mean2 = df.sample(1)\n",
    "    for G4 in range(start,stop,inc):\n",
    "        i = 0\n",
    "        try: \n",
    "            while i < permutation: \n",
    "                df_now = df[(df[\"G4_Count\"] == G4)].sample(sample).mean().reset_index().T\n",
    "                new_header = df_now.iloc[0] #grab the first row for the header\n",
    "                df_now = df_now[1:] #take the data less the header row\n",
    "                df_now.columns = new_header #set the header row as the df header\n",
    "                df_mean2 = pd.concat([df_mean2, df_now])\n",
    "\n",
    "                i += 1\n",
    "        except:\n",
    "            continue\n",
    "    \n",
    "    y1 = df_mean2[key1].fillna(0).values\n",
    "    y2 = df_mean2[key2].fillna(0).values\n",
    "    X = df_mean2['G4_Count'].fillna(0).values.reshape(-1,1)\n",
    "\n",
    "    ols1 = linear_model.LinearRegression()\n",
    "    model1 = ols1.fit(X, y1)\n",
    "    r2_1 = model1.score(X, y1)\n",
    "    \n",
    "    ols2 = linear_model.LinearRegression()\n",
    "    model2 = ols2.fit(X, y2)\n",
    "    r2_2 = model2.score(X, y2)\n",
    "    \n",
    "    \n",
    "    df_now = pd.DataFrame({\"Sample_Size\": [sample],\n",
    "                           \"R2_1\": [r2_1],\n",
    "                           \"R2_2\": [r2_2],\n",
    "                           \"Coef_1\": [model1.coef_],\n",
    "                           \"Coef_2\": [model2.coef_]})\n",
    "    df_results = pd.concat([df_results, df_now])\n",
    "\n",
    "# create figure and axis objects with subplots()\n",
    "\n",
    "fig,ax = plt.subplots()\n",
    "\n",
    "fig.set_size_inches(6, 6)\n",
    "# make a plot\n",
    "ax.plot(df_results[\"Sample_Size\"], df_results[\"R2_2\"], color=cSNPs, marker=\"o\")\n",
    "\n",
    "ax.plot(df_results[\"Sample_Size\"], df_results[\"R2_1\"],color=SNPs,marker=\"o\")\n",
    "ax.set_ylim([-.05, 1])\n",
    "\n",
    "\n",
    "for pos in ['right', 'top']:\n",
    "    plt.gca().spines[pos].set_visible(False)\n",
    "ax.set_xlabel('Sample Size')\n",
    "plt.rcParams.update({\n",
    "    'font.size': 15,         # default font size\n",
    "    'axes.labelsize': 15,    # x and y labels\n",
    "    'xtick.labelsize':15,   # x-axis tick labels\n",
    "    'ytick.labelsize': 15,   # y-axis tick labels\n",
    "    'legend.fontsize': 25,   # legend font size\n",
    "    'figure.titlesize': 40,  # figure title font size\n",
    "})\n",
    "ax.set_ylabel(\"R2 of linear fit\",fontsize=25)\n",
    "ax.set_xlabel(\"Sample size\",fontsize=25)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ac9d5c9-0bbb-4d6a-8ba3-f93a0c2a737c",
   "metadata": {},
   "source": [
    "## Plot incomplete categorical bootstrapping for the full dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c95d76ce-25f5-4876-a9cc-2e7149ded16f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "key1 = \"SNP_Count_normalized\"#_Variance_normalized\"\n",
    "key2 = \"cSNP_Count_normalized\"#_Variance_normalized\"\n",
    "df_results = pd.DataFrame({\"Sample_Size\": [],\n",
    "                           \"R2_1\": [],\n",
    "                           \"R2_2\": [],\n",
    "                           \"Coef_1\": [],\n",
    "                           \"Coef_2\": []})\n",
    "start = 0\n",
    "stop = 16\n",
    "inc = 1\n",
    "permutation = 250\n",
    "c = cSNPs\n",
    "figure(figsize=(5,5), dpi=500)\n",
    "for sample in range(1,102,5):\n",
    "    data = []\n",
    "    df_mean2 = df.sample(1)\n",
    "\n",
    "\n",
    "    for G4 in range(start,stop,inc):\n",
    "        i = 0\n",
    "        try: \n",
    "            while i < permutation: \n",
    "                df_now = df[(df[\"G4_Count\"] == G4)].sample(sample).mean().reset_index().T\n",
    "                new_header = df_now.iloc[0] #grab the first row for the header\n",
    "                df_now = df_now[1:] #take the data less the header row\n",
    "                df_now.columns = new_header #set the header row as the df header\n",
    "                df_mean2 = pd.concat([df_mean2, df_now])\n",
    "\n",
    "                i += 1\n",
    "        except:\n",
    "            continue\n",
    "    \n",
    "    #y1 = df_mean2[key1].fillna(0).values\n",
    "    y2 = df_mean2[key2].fillna(0).values\n",
    "    X = df_mean2['G4_Count'].fillna(0).values.reshape(-1,1)\n",
    "\n",
    "    #ols1 = linear_model.LinearRegression()\n",
    "    #model1 = ols1.fit(X, y1)\n",
    "    #r2_1 = model1.score(X, y1)\n",
    "    \n",
    "    ols2 = linear_model.LinearRegression()\n",
    "    model2 = ols2.fit(X, y2)\n",
    "    r2_2 = model2.score(X, y2)\n",
    "    \n",
    "    \n",
    "    df_now = pd.DataFrame({\"Sample_Size\": [sample],\n",
    "                           #\"R2_1\": [r2_1],\n",
    "                           \"R2_2\": [r2_2],\n",
    "                           #\"Coef_1\": [model1.coef_],\n",
    "                           \"Coef_2\": [model2.coef_]})\n",
    "    df_results = pd.concat([df_results, df_now])\n",
    "\n",
    "    for G4 in range(start, stop, inc): \n",
    "        data.append(df_mean2[df_mean2[\"G4_Count\"] == G4][key2])\n",
    "    boxplot_dict = plt.boxplot(data, showfliers=False, whis=0, showcaps=False, patch_artist=True,\n",
    "            medianprops=dict(color=\"black\"),)\n",
    "    for b in boxplot_dict['boxes']:\n",
    "        b.set_alpha(0.1)\n",
    "        b.set(color='black', linewidth=1)\n",
    "        b.set(facecolor = cSNPs)\n",
    "for pos in ['right', 'top']:\n",
    "    plt.gca().spines[pos].set_visible(False)\n",
    "plt.rcParams.update({\n",
    "    'font.size': 15,         # default font size\n",
    "    'axes.labelsize': 15,    # x and y labels\n",
    "    'xtick.labelsize':15,   # x-axis tick labels\n",
    "    'ytick.labelsize': 15,   # y-axis tick labels\n",
    "    'legend.fontsize': 25,   # legend font size\n",
    "    'figure.titlesize': 40,  # figure title font size\n",
    "})\n",
    "plt.ylabel(\"Normalized cSNPs / 5kb\",fontsize=15)\n",
    "plt.xlabel('G4s / 5kb',fontsize=15)\n",
    "plt.savefig(\"output_bootstrap_2_plot.png\", dpi=500)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79c3027a-5026-41d6-92f3-983c5dd7aa72",
   "metadata": {},
   "source": [
    "## Comparing the average number of normalized cSNVs across different cancer types in genomic windows containing 0/>=10 G4s. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45e8552c-52e1-4441-bfa0-6cb5a92b1f75",
   "metadata": {},
   "outputs": [],
   "source": [
    "test1 = df_cancer[df_cancer[\"Patient_Count\"] > 14][\"0_n\"]\n",
    "test2 = df_cancer[df_cancer[\"Patient_Count\"] > 14][\"10_n\"]\n",
    "fig = go.Figure()\n",
    "fig.add_trace(go.Box(\n",
    "    y=test1,\n",
    "    name='No G4s',\n",
    "    boxmean=True,\n",
    "    line_color='black',\n",
    "    fillcolor=\"lightsteelblue\",\n",
    "    opacity=0.6,\n",
    "    boxpoints=\"all\"\n",
    "))\n",
    "fig.add_trace(go.Box(\n",
    "    y=test2,\n",
    "    name='>= 10 G4s',\n",
    "    boxmean=True,\n",
    "    boxpoints=\"all\",\n",
    "    line_color='black',\n",
    "    fillcolor=\"lightcoral\",\n",
    "    opacity=0.6,\n",
    "))\n",
    "\n",
    "fig.update_layout(\n",
    "    autosize=False,\n",
    "    width=600,\n",
    "    height=600,\n",
    "    margin=dict(\n",
    "        l=50,\n",
    "        r=50,\n",
    "        b=100,\n",
    "        t=100,\n",
    "        pad=4\n",
    "    ),\n",
    ")\n",
    "fig.update_layout(\n",
    "    #title=\"Plot Title\",\n",
    "    yaxis_title=\"Normalized SNPs / 5 kb, grouped for cancer types\",\n",
    "    #legend_title=\"Legend Title\",\n",
    "    )\n",
    "fig.for_each_trace(lambda t: t.update({\"marker\":{\"size\":3}}))\n",
    "#fig.update_yaxes(type=\"log\")\n",
    "fig.update_yaxes(title=\"Cancer Types, normalized cSNPs / 5kb\")\n",
    "fig.update_layout(\n",
    "    yaxis_range=[0, 70],\n",
    "    template=\"simple_white\",\n",
    "    #legend_title=\"Legend Title\",\n",
    "    yaxis = dict(tickfont = dict(size=25)),\n",
    "    xaxis = dict(tickfont = dict(size=20)),\n",
    "    #legend = dict(font = dict(size = 20, color = \"black\")),\n",
    "    font=dict(size=30))\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "189dceaa-df8f-4bc3-8ea0-ca7b7ab370b0",
   "metadata": {},
   "source": [
    "## Group patients by cancer type and tissue and calculate median G4-averages across patients with similiar diagnoses. Plot as sunburst plot. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37376173-3664-4499-84e6-4f5526ba0460",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "key = \"G4_n\"\n",
    "df_results_normalized = pd.DataFrame({\"primary_diagnosis\": [],\n",
    "                           \"tissue\": [],\n",
    "                           \"Patient_Count\": [],\n",
    "                           \"G4_mean\": [],\n",
    "                           \"G4_median\": [],\n",
    "                           \"G4_var\": [],\n",
    "                           \"G4_std\": []})\n",
    "\n",
    "diagnoses = df_m3.primary_diagnosis.unique()\n",
    "\n",
    "length = len(diagnoses)\n",
    "i = 0\n",
    "for diagnosis in diagnoses:\n",
    "    tissues = df_m3[df_m3[\"primary_diagnosis\"] == diagnosis][\"tissue_or_organ_of_origin\"].unique()\n",
    "    for tissue in tissues:\n",
    "        sub = df_m3[(df_m3[\"primary_diagnosis\"] == diagnosis) & (df_m3[\"tissue_or_organ_of_origin\"] == tissue)]\n",
    "        sub[\"G4_n\"] = sub[\"G4_Count\"] * (sub[\"G4_Count\"] / sub[\"G4_Count\"].sum()) \n",
    "        df_now = pd.DataFrame({\"primary_diagnosis\": [diagnosis],\n",
    "                               \"tissue\": [tissue],\n",
    "                               \"Patient_Count\": [len(sub[\"case_id\"].unique())],\n",
    "                               \"G4_mean\": [sub[key].mean()],\n",
    "                               \"G4_median\": [sub[key].median()],\n",
    "                               \"G4_var\": [sub[key].var()],\n",
    "                               \"G4_std\": [sub[key].std()]})\n",
    "        df_results_normalized = pd.concat([df_results_normalized, df_now])\n",
    "    i += 1\n",
    "    print(i/length)\n",
    "\n",
    "df_results_grouped_normalized = pd.DataFrame({\"primary_diagnosis\": [],\n",
    "                           \"tissue\": [],\n",
    "                           \"Patient_Count\": [],\n",
    "                           \"G4_mean\": [],\n",
    "                           \"G4_median\": [],\n",
    "                           \"G4_var\": [],\n",
    "                           \"G4_std\": []})\n",
    "\n",
    "diagnoses = df_m3.primary_diagnosis.unique()\n",
    "\n",
    "length = len(diagnoses)\n",
    "i = 0\n",
    "for diagnosis in diagnoses:\n",
    "    tissues = df_m3[df_m3[\"primary_diagnosis\"] == diagnosis][\"tissue_or_organ_of_origin\"].unique()\n",
    "    for tissue in tissues:\n",
    "        sub = df_m3[(df_m3[\"primary_diagnosis\"] == diagnosis) & (df_m3[\"tissue_or_organ_of_origin\"] == tissue)]\n",
    "        sub[\"G4_n\"] = sub[\"G4_Count\"] * (sub[\"G4_Count\"] / sub[\"G4_Count\"].sum())\n",
    "        df_now = pd.DataFrame({\"primary_diagnosis\": [diagnosis],\n",
    "                               \"tissue\": [tissue],\n",
    "                               \"Patient_Count\": [len(sub[\"case_id\"].unique())],\n",
    "                               \"G4_mean\": [sub.groupby([\"case_id\"])[key].mean().mean()],\n",
    "                               \"G4_median\": [sub.groupby([\"case_id\"])[key].mean().median()],\n",
    "                               \"G4_var\": [sub.groupby([\"case_id\"])[key].mean().var()],\n",
    "                               \"G4_std\": [sub.groupby([\"case_id\"])[key].mean().std()]})\n",
    "        df_results_grouped_normalized = pd.concat([df_results_grouped_normalized, df_now])\n",
    "    i += 1\n",
    "    print(i/length)\n",
    "    \n",
    "df_results_normalized[\"log_mean\"] = np.log10(df_results_normalized[\"G4_mean\"])\n",
    "df_results_grouped_normalized[\"log_std\"] = np.log10(df_results_normalized[\"G4_std\"] / df_results_normalized[\"G4_mean\"])\n",
    "df_results_grouped_normalized[\"norm_std\"] = df_results_normalized[\"G4_std\"] / df_results_normalized[\"G4_mean\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "354c16d8-3d98-4ed4-b6c6-362d32d993c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "key = \"log_mean\"\n",
    "sub = df_results_normalized[df_results_normalized[\"Patient_Count\"] > 200].sort_values(by=key)\n",
    "sub[key] = sub[key]# * 100000\n",
    "\n",
    "fig = px.sunburst(\n",
    "    sub,\n",
    "    path=[\"tissue\", \"primary_diagnosis\"],\n",
    "    color=key,\n",
    "    color_continuous_scale='RdBu',\n",
    "    #range_color=[min(sub[key]), max(sub[key])],\n",
    "    color_continuous_midpoint=np.average(sub[key]),# weights=sub[key]),\n",
    "    width = 800,\n",
    "    height = 800)\n",
    "\n",
    "fig.update_layout(\n",
    "    template=\"simple_white\",\n",
    "    #legend_title=\"Legend Title\",\n",
    "    yaxis = dict(tickfont = dict(size=25)),\n",
    "    xaxis = dict(tickfont = dict(size=25)),\n",
    "    #legend = dict(font = dict(size = 20, color = \"black\")),\n",
    "    font=dict(size=20))\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3014be75-2ba6-45b9-bc05-892700ccc06d",
   "metadata": {},
   "source": [
    "## Group patients by cancer type and calculate intra-group mean-normalized standard deviation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2d08875-ee07-42da-8eb2-e04af1ca52d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sub = df_results_grouped_normalized[(df_results_grouped_normalized[\"Patient_Count\"] > 50)].groupby([\"primary_diagnosis\"]).mean().reset_index().sort_values(by=\"norm_std\", ascending=False)\n",
    "# Generate some example data\n",
    "np.random.seed(42)\n",
    "scatter_x = sub[\"temp\"]\n",
    "scatter_y = sub[\"norm_std\"]\n",
    "box_data = sub[\"norm_std\"]\n",
    "new_min = 10\n",
    "new_max = 30\n",
    "min_value = sub['Patient_Count'].min()\n",
    "max_value = sub['Patient_Count'].max()\n",
    "\n",
    "sub['Normalized_Patient_Count'] = (sub['Patient_Count'] - min_value) * (new_max - new_min) / (max_value - min_value) + new_min\n",
    "size = sub['Normalized_Patient_Count']\n",
    "# Create the scatter plot\n",
    "scatter = go.Scatter(\n",
    "    x=scatter_x,\n",
    "    y=scatter_y,\n",
    "    mode='markers',\n",
    "    marker=dict(\n",
    "        size=size,\n",
    "        color='rgba(50, 50, 50, 0.8)'\n",
    "    ),\n",
    "    name='Scatter Plot'\n",
    ")\n",
    "\n",
    "# Create the box plot\n",
    "box = go.Box(\n",
    "    y=box_data,\n",
    "    x=scatter_x,\n",
    "    fillcolor = 'rgba(0, 0, 0, 0.1)',\n",
    "    name='Box Plot',\n",
    "    marker=dict(\n",
    "        \n",
    "        color='rgba(0, 0, 0, 0.8)'\n",
    "    ),\n",
    "    line=dict(\n",
    "        color='rgba(0, 0, 0, 0.4)'\n",
    "    ),\n",
    "    boxpoints=False,\n",
    "    notched=True# Only show outlier points\n",
    "    #jitter=0.3,  # Add some jitter for better visibility of points\n",
    "    #pointpos=0  # Position of the points relative to the box\n",
    ")\n",
    "\n",
    "# Create the layout for the chart\n",
    "layout = go.Layout(\n",
    "    #title='Overlapping Scatter and Box Plots',\n",
    "    #xaxis=dict(title='X-axis Label'),\n",
    "    yaxis=dict(title='Normalized std'),\n",
    "    height=500,\n",
    "    width=500\n",
    ")\n",
    "\n",
    "# Create the figure and add the scatter and box plots\n",
    "fig = go.Figure(data=[scatter, box], layout=layout)\n",
    "fig.update_layout(\n",
    "    template=\"simple_white\",\n",
    "    #legend_title=\"Legend Title\",\n",
    "    yaxis = dict(tickfont = dict(size=25)),\n",
    "    xaxis = dict(tickfont = dict(size=20)),\n",
    "    #legend = dict(font = dict(size = 20, color = \"black\")),\n",
    "    font=dict(size=30))\n",
    "# Show the figure\n",
    "fig.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09bfdc1a-84f1-4dd2-9934-a1e77be8657f",
   "metadata": {},
   "source": [
    "## Fit Cox's proportional hazards model to patients that are eeither alive or died due to cancer-related circumstances. Use cSNV-average and G4-average as potential predictor variables.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7fb2ebd-6663-4e5e-b009-977a5c6594a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_c = pd.read_csv(\"/mnt/c/Users/tilma/Desktop/WORK/SNPs/SNP_DATA/GDC_Database/clinical.tsv\", sep=\"\\t\")\n",
    "df_closest = pd.read_csv(\"ALL_GDC_exons_5kb_windows_G4s.closest.temp\", sep=\"\\t\", header=None).rename(columns={0: \"SNP_ID\", 1: \"Distance\"})\n",
    "temp = pd.merge(df_m3[(df_m3[\"cause_of_death\"] == \"Cancer Related\") | (df_m3[\"cause_of_death\"] == \"'--\") & (df_m3[\"vital_status\"] != \"Not Reported\")][[\"SNP_ID\", \"G4_Count\", \"case_id\", \"SNP_Count\"]], df_closest, on=\"SNP_ID\", how=\"left\")\n",
    "temp = temp.drop_duplicates(subset=\"SNP_ID\", keep=\"first\")\n",
    "temp[\"Distance\"] = temp[\"Distance\"].abs()\n",
    "data2 = pd.merge(temp, df_c[[\"case_id\", \"days_to_death\", \"cause_of_death\"]], on=\"case_id\", how=\"left\").drop_duplicates(keep=\"first\")\n",
    "#data2 = pd.merge(df_m3[[\"G4_Count\", \"case_id\", \"SNP_Count\"]].drop_duplicates(keep=\"first\"), df_c[[\"case_id\", \"days_to_death\", \"vital_status\"]], on=\"case_id\", how=\"left\")\n",
    "#data2['days_to_death'] = data2['days_to_death'].replace(\"'--\", default)\n",
    "df_cm = df_c\n",
    "#df_cm['days_to_death'] = df_cm['days_to_death'].replace(\"'--\", default)\n",
    "data3 = pd.merge(data2[[\"G4_Count\", \"case_id\", \"days_to_death\", \"Distance\", \"SNP_Count\"]].groupby([\"case_id\"]).median().reset_index(), df_cm[pd.to_numeric(df_cm['days_to_death'], errors='coerce').notnull()][[\"case_id\", \"days_to_death\", \"cause_of_death\"]], on=\"case_id\", how=\"left\").drop_duplicates(subset=\"case_id\", keep=\"first\")\n",
    "data4 = pd.merge(data2[[\"G4_Count\", \"case_id\", \"days_to_death\", \"Distance\"]].groupby([\"case_id\"]).var().reset_index().rename(columns={\"G4_Count\": \"G4_Count_Var\", \"Distance\": \"Distance_Var\"}), data3, on=\"case_id\", how=\"left\").dropna()\n",
    "\n",
    "# Assuming your preprocessed DataFrame is named 'data'\n",
    "\n",
    "data = data4#[data4[\"Distance\"] < 100]#.drop_duplicates(subset=\"case_id\", keep=\"first\")#.sample(10000)\n",
    "#data = data2.sample(10000)\n",
    "data = data[pd.to_numeric(data['days_to_death'], errors='coerce').notnull()]\n",
    "data['event'] = data['cause_of_death'].apply(lambda x: 1 if x == \"Cancer Related\" else 0)\n",
    "# Keep a copy of the data with primary_diagnosis for step 5\n",
    "#data_with_diagnosis = data.copy()\n",
    "\n",
    "\n",
    "# Remove non-numeric columns\n",
    "#data = data.drop(['vital_status', 'primary_diagnosis'], axis=1)\n",
    "data = data[[\"event\", \"days_to_death\", \"G4_Count\", \"SNP_Count\"]]\n",
    "#data = data[[\"event\", \"days_to_death\", \"G4_Count\", \"Distance\"]]\n",
    "# Fit the Cox proportional hazards model\n",
    "cph = CoxPHFitter()\n",
    "cph.fit(data, duration_col='days_to_death', event_col='event')\n",
    "\n",
    "# Print the summary\n",
    "cph.print_summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad1fe00f-be9d-40cd-8480-32ebcb720d55",
   "metadata": {},
   "source": [
    "## Plot results of Cox's proportional hazards model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0937fee1-15ca-4590-b84c-aeffde95b7b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "x_labels = ['SNP Average', 'G4 Average']\n",
    "coef_values = [1.0, 1.46]\n",
    "lower_ci = [0.99, 1.28]\n",
    "upper_ci = [1.01, 1.67]\n",
    "\n",
    "bar_colors = ['blue', 'orange']  # Choose the colors for the bars\n",
    "\n",
    "# Create the bar plot\n",
    "fig, ax = plt.subplots(figsize=(8, 3))\n",
    "bars = ax.barh(x_labels, coef_values, xerr=errors, capsize=5, color=bar_colors)\n",
    "\n",
    "# Set the background color to white\n",
    "ax.set_facecolor('white')\n",
    "fig.set_facecolor('white')\n",
    "\n",
    "# Leave the left and bottom grid axes\n",
    "ax.xaxis.grid(True, which='both', linestyle='--', linewidth=0.5)\n",
    "ax.yaxis.grid(False)\n",
    "for pos in ['right', 'top']:\n",
    "    plt.gca().spines[pos].set_visible(False)\n",
    "plt.rcParams.update({\n",
    "    'font.size': 15,         # default font size\n",
    "    'axes.labelsize': 15,    # x and y labels\n",
    "    'xtick.labelsize':15,   # x-axis tick labels\n",
    "    'ytick.labelsize': 15,   # y-axis tick labels\n",
    "    'legend.fontsize': 25,   # legend font size\n",
    "    'figure.titlesize': 40,  # figure title font size\n",
    "})\n",
    "plt.ylabel(\"Normalized cSNPs / 5kb\",fontsize=15)\n",
    "plt.xlabel('G4s / 5kb',fontsize=15)\n",
    "plt.savefig(\"output_Cox_plot.png\", dpi=500)\n",
    "# Display the plot\n",
    "plt.show()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
